src/probe/probe_iterator.jl

```diff
@@ -57,19 +67,73 @@ function probe(examples::Vector{<:IOExample}, iterator::ProgramIterator, select:
 
     return nothing
 end
+"""
+    selectpsol_largest_subset(partial_sols::Vector{ProgramCache}) 
 
-@programiterator GuidedSearchIterator(
-    spec::Vector{<:IOExample},
-    symboltable::SymbolTable
-)
+This scheme selects a single cheapest program (first enumerated) that 
+satisfies the largest subset of examples encountered so far across all partial_sols.
+"""
+function selectpsol_largest_subset(partial_sols::Vector{ProgramCache})  
+    if isempty(partial_sols)
+        return []
+    end
+    largest_subset_length = maximum(cache -> length(cache.correct_examples), partial_sols)
+    programs_with_largest_length = filter(cache -> length(cache.correct_examples) == largest_subset_length, partial_sols)
+    # find the program with the smallest cost
+    minimum_cost = minimum(cache -> cache.cost, programs_with_largest_length)
+    best_sol = first(filter(cache -> cache.cost == minimum_cost, programs_with_largest_length))
+    return [best_sol]
+end
 
-@kwdef mutable struct GuidedSearchState 
-    level::Int64
-    bank::Vector{Vector{RuleNode}}
-    eval_cache::Set
-    programs::Vector{RuleNode}
+"""
+    selectpsol_first_cheapest(partial_sols::Vector{ProgramCache}) 
+
+This scheme selects a single cheapest program (first enumerated) that 
+satisfies a unique subset of examples.
+"""
+function selectpsol_first_cheapest(partial_sols::Vector{ProgramCache})  
+    # maps subset of examples to the cheapest program 
+    mapping = Dict{Vector{Int}, ProgramCache}()
+    for sol ∈ partial_sols
+        examples = sol.correct_examples
+        if !haskey(mapping, examples)
+            mapping[examples] = sol
+        else
+            # if the cost of the new program is less than the cost of the previous program with the same subset of examples replace it
+            if sol.cost < mapping[examples].cost
+                mapping[examples] = sol
+            end
+        end
+    end
+    # get the cheapest programs that satisfy unique subsets of examples
+    return values(mapping)
 end
 
+"""
+    selectpsol_all_cheapest(partial_sols::Vector{ProgramCache}) 
+
+This scheme selects all cheapest programs that satisfies a unique subset of examples.
+"""
+function selectpsol_all_cheapest(partial_sols::Vector{ProgramCache})  
+    # maps subset of examples to the cheapest program 
+    mapping = Dict{Vector{Int}, Vector{ProgramCache}}()
+    for sol ∈ partial_sols
+        examples = sol.correct_examples
+        if !haskey(mapping, examples)
+            mapping[examples] = [sol]
+        else
+            # if the cost of the new program is less than the cost of the first program
+            if sol.cost < mapping[examples][begin].cost
+                mapping[examples] = [sol]
+            elseif sol.cost == mapping[examples][begin].cost
+                # append to the list of cheapest programs
+                push!(mapping[examples], sol)
+            end
```

Comment:
I think it would be better to save `mapping[examples]` in a variable, so that it does not have to find it three times (worst case).
```suggestion
            progs = mapping[examples]
            if sol.cost < progs[begin].cost
                mapping[examples] = [sol]
            elseif sol.cost == progs[begin].cost
                # append to the list of cheapest programs
                push!(progs, sol)
            end
```